### PR TITLE
Use ReturnCode throughout kernel

### PR DESCRIPTION
**Pull request by ppannuto**
The driver trait now enforces sanctioned return codes throughout the kernel. Drivers can return any positive (non-error) value using `ReturnCode::SuccessWithValue`

I believe that all of the integer return codes were changed into the appropriate return code, but driver authors should double-check

### CODE REVIEW

**Comment by alevy** [comment]
You are a true hero, @ppannuto

**Comment by ppannuto** [comment]
Heads up if you pulled that branch, I just force pushed a fix to the second patch for rustfmt

File: capsules/src/console.rs [hunk 1]
```diff
@@ -127,27 +141,35 @@ impl<'a, U: UART> Driver for Console<'a, U> {
                                 app.pending_write = true;
                                 app.write_buffer = Some(slice);
                             }
-                            0
+                            ReturnCode::SUCCESS
                         },
-                        None => -1
+                        None => ReturnCode::FAIL
+                        // XXX  ^^^^^^^^^^^^^^^^-- When can we fail to take the write_buffer
+                        //                         which return code is best here?
```
Comment (ppannuto) [inline capsules/src/console.rs]: I believe this error case is when a user calls subscribe without first calling allow to populate that buffer, right? This is sort of the like `ERESERVE`, but not quite. Thoughts?

File: capsules/src/nrf51822_serialization.rs [hunk 2]
```diff
@@ -136,24 +137,27 @@ impl<'a, U: UARTAdvanced> Driver for Nrf51822Serialization<'a, U> {
...
 
             // Send a buffer to the nRF51822 over UART.
             1 => {
                 // On a TX, send the first byte of the TX buffer.
                 // TODO(bradjc): Need to match this to the correct app!
                 //               Can't just use 0!
-                let result = self.app.map(|appst| {
+                //
+                // When could app be NULL? What return code should be here?
+                // XXX ReturnCode -----------,,,,,,,,,,,,,,,,
```
Comment (ppannuto) [inline capsules/src/nrf51822_serialization.rs]: cc @bradjc

File: capsules/src/nrf51822_serialization.rs [hunk 3]
```diff
@@ -164,12 +168,13 @@ impl<'a, U: UARTAdvanced> Driver for Nrf51822Serialization<'a, U> {
                                 }
                                 self.uart.transmit(buffer, write_len);
                             });
-                            0
+                            ReturnCode::SUCCESS
                         }
-                        None => -2,
+                        None => ReturnCode::FAIL, /* XXX: ReturnCode */
```
Comment (ppannuto) [inline capsules/src/nrf51822_serialization.rs]: cc @bradjc -- maybe the same case / error code as the console comment above

File: capsules/src/spi.rs [hunk 4]
```diff
@@ -186,19 +187,17 @@ impl<'a, S: SpiMasterDevice> Driver for Spi<'a, S> {
...
-            1 /* read_write_byte */ => -1,
+            1 /* read_write_byte */ => ReturnCode::ENOSUPPORT,
             2 /* read_write_bytes */ => {
                 if self.busy.get() {
-                    return -1;
+                    return ReturnCode::EBUSY;
                 }
-                let mut result = -1;
-                self.app.map(|app| {
+                self.app.map_or(ReturnCode::FAIL /* XXX app is null? */, |app| {
```
Comment (ppannuto) [inline capsules/src/spi.rs]: I think this is the same issue as console and nrfSerialization (missing allow first)? cc @phil-levis

File: capsules/src/spi.rs [hunk 5]
```diff
@@ -210,46 +209,51 @@ impl<'a, S: SpiMasterDevice> Driver for Spi<'a, S> {
...
-                   // so virtual instances can use multiple chip selects
+                // do nothing, for now, until we fix interface
+                // so virtual instances can use multiple chip selects
+                ReturnCode::ENOSUPPORT
             }
             4 /* get chip select */ => {
-                0
+                //XXX Was a naked, uncommented zero. I'm assuming that's
+                //    because the only valid chip select for now is 0,
+                //    and wrapping it appropriately -Pat, ReturnCode fixes
```
Comment (ppannuto) [inline capsules/src/spi.rs]: cc @phil-levis

**Comment by ppannuto** [comment]
I'm going to merge this as it touches a lot of code and the lingering issues aren't incorrect, just imprecise. I'll open an issue for the remaining nits that I identified.

**Pull request merged**